### PR TITLE
VLTCLT-42: support for quotas apis

### PR DESCRIPTION
**Pull request by benzekrimaha**
This pull request includes changes to the `VaultClient` class in the `lib/IAMClient.js` file. The main changes are to the method signatures and the handling of `accountName` in the `updateAccountQuota`, `deleteAccountQuota`, and the addition of the`getAccountQuota` method. 
The `accountName` parameter is now passed as part of an `options` object, which allows for more flexibility and extensibility in the future.
Additionally, the `quota` validation in the `updateAccountQuota` method has been simplified. 
Jira issue here : https://scality.atlassian.net/browse/VLTCLT-42

When working on the ticket we discovered that the errors sent to vault needed to be restructured : 
- Changed the error structure to match with the expected format on vault2 side 
- Added properties to the InternalError object to allow for more specific error identification.
- Updated the test case to directly compare the error code and description with the InternalError object.

### CODE REVIEW

File: lib/constants.js [hunk 1]
```diff
@@ -3,6 +3,10 @@ const httpClientFreeSocketTimeout = 55000;
 const InternalError = {
     code: 500,
     description: 'We encountered an internal error. Please try again.',
+    InternalError: true,
+    is: {
+        InternalError: true
+    }
```
Comment (ghost) [inline lib/constants.js]:
```suggestion
    is: {
        InternalError: true,
    },
```

File: lib/IAMClient.js [hunk 2]
```diff
@@ -1207,6 +1244,10 @@ class VaultClient {
                 return cb({
                         code: obj.ErrorResponse.Error.Code,
                         description: obj.ErrorResponse.Error.Message,
+                        [obj.ErrorResponse.Error.Code]: true,
+                        is: {
+                        [obj.ErrorResponse.Error.Code]: true,
+                        },
```
Comment (ghost) [inline lib/IAMClient.js]:
```suggestion
                        is: {
                            [obj.ErrorResponse.Error.Code]: true,
                        },
```

File: bin/vaultclient [hunk 3]
```diff
@@ -213,20 +213,25 @@ program
 
 program
     .command('update-account-quota')
-    .option('--account-name <NAME>', 'Name of the account')
     .option('--quota <QUOTA>', 'Maximum quota for the account', parseInt)
+    .option('--account-name <NAME>', 'Name of the account')
     .action(action.bind(null, 'update-account-quota', (client, args) => {
-        client.updateAccountQuota(args.accountName, args.quota,
-            handleVaultResponse);
+        client.updateAccountQuota(args.quota, handleVaultResponse, { accountName: args.accountName || undefined});
```
Comment (ghost) [inline bin/vaultclient]:
```suggestion
        client.updateAccountQuota(args.quota, handleVaultResponse, { accountName: args.accountName || undefined });
```

Do we really need the `undefined` part? Can't that be handled in the function directly?

File: bin/vaultclient [hunk 4]
```diff
@@ -213,20 +213,25 @@ program
...
     }));
 
 program
     .command('delete-account-quota')
     .option('--account-name <NAME>', 'Name of the account')
     .action(action.bind(null, 'delete-account-quota', (client, args) => {
-        client.deleteAccountQuota(args.accountName, handleVaultResponse);
+        client.deleteAccountQuota(handleVaultResponse, { accountName: args.accountName || undefined});
```
Comment (ghost) [inline bin/vaultclient]:
```suggestion
        client.deleteAccountQuota(handleVaultResponse, { accountName: args.accountName || undefined });
```

File: bin/vaultclient [hunk 5]
```diff
@@ -213,20 +213,25 @@ program
...
     }));
 
+program
+    .command('get-account-quota')
+    .option('--account-name <NAME>', 'Name of the account')
+    .action(action.bind(null, 'get-account-quota', (client, args) => {
+        client.getAccountQuota( handleVaultResponse, {accountName: args.accountName || undefined});
```
Comment (ghost) [inline bin/vaultclient]:
```suggestion
        client.getAccountQuota( handleVaultResponse, {accountName: args.accountName || undefined });
```

File: lib/constants.js (hunk 1 above)
Comment (ghost) [reply lib/constants.js]: I wonder if we really need the `is`; for the error name, it makes sense as it is standard, but for the `is`, it is only something we rely with arsenal errors...

File: lib/IAMClient.js [hunk 6]
```diff
@@ -284,40 +284,77 @@ class VaultClient {
...
+    getAccountQuota(callback, options) {
+        const data = {
+            Action: 'GetAccountQuota',
+            Version: '2010-05-08',
+        };
+        if (options) {
+            const { accountName } = options;
+            if (accountName) {
+                assert(typeof accountName === 'string' && accountName !== '',
+                    'name should be a string');
```
Comment (ghost) [inline lib/IAMClient.js]: As this code is under `if (accountName)`, the `accountName !== ''` check is always true

Comment (ghost) [reply lib/IAMClient.js]:
```suggestion
                assert(typeof accountName === 'string',
                    'the account name, if set, should be a string');
```

File: lib/IAMClient.js
Comment (ghost) [inline lib/IAMClient.js:1]: I think, even if that's not done today, that we should test these APIs in our unit tests

Comment (benzekrimaha) [reply lib/IAMClient.js:1]: Done here : https://github.com/scality/vaultclient/pull/442/commits/b778f49301a3ac75e7d7535a3a6e8c9de2478f99

File: lib/IAMClient.js (hunk 6 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: Done here : https://github.com/scality/vaultclient/pull/442/commits/b778f49301a3ac75e7d7535a3a6e8c9de2478f99#diff-86a8509a2ad22b19f93123bda4efc46cfa71b4c2bbfded93e5f6e1a25e69f180R338

File: bin/vaultclient (hunk 5 above)
Comment (benzekrimaha) [reply bin/vaultclient]: Keeping for consistency with other tests

File: bin/vaultclient (hunk 3 above)
Comment (benzekrimaha) [reply bin/vaultclient]: Keeping this for consistency with other tests

File: lib/constants.js (hunk 1 above)
Comment (benzekrimaha) [reply lib/constants.js]: Has been removed here : https://github.com/scality/vaultclient/pull/442/commits/c9da2fd343b2396ad6350a96063fc6fa26d5d16b#diff-602f56bb692d9bf3bdc5fef3d75d5fa33fe841c3d21ca46294bb6fdaf0785d0cL7-L9

File: lib/IAMClient.js (hunk 2 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: Has been removed here : https://github.com/scality/vaultclient/pull/442/commits/c9da2fd343b2396ad6350a96063fc6fa26d5d16b#diff-86a8509a2ad22b19f93123bda4efc46cfa71b4c2bbfded93e5f6e1a25e69f180L1262-L1264

File: bin/vaultclient (hunk 4 above)
Comment (benzekrimaha) [reply bin/vaultclient]: done : https://github.com/scality/vaultclient/pull/442/commits/8e839aa1ed22f5bb3a647feb5b4a00c386b721d5#diff-6ed3f39879bec9944e1a6e9d0d3c1bcbc7209a7eab826e5c3819048e4841163dR226

File: bin/vaultclient (hunk 5 above)
Comment (benzekrimaha) [reply bin/vaultclient]: done : https://github.com/scality/vaultclient/pull/442/commits/8e839aa1ed22f5bb3a647feb5b4a00c386b721d5#diff-6ed3f39879bec9944e1a6e9d0d3c1bcbc7209a7eab826e5c3819048e4841163dR233

File: lib/IAMClient.js [hunk 7]
```diff
@@ -284,40 +284,90 @@ class VaultClient {
...
+        this.request('POST', '/', true, callback, data);
     }
 
     /**
      * Delete Quota of an account
      *
-     * @param {string} accountName - account name
+     * @param {object} options - additional creation params
+     * @param {string} [options.accountName] - account name
+     * @param {boolean} [options.ci] - test flag
```
Comment (ghost) [inline lib/IAMClient.js]:
```suggestion
     * @param {boolean} [options.ci] - boolean used for testing. When
     *  set to true, it helps sending invalid requests, to ensure
     *  input validation.
```

File: tests/unit/updateAccountQuota.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+        client.updateAccountQuota(quota, undefined, () => {});
+        assert.strictEqual(lastRequestData.method, 'POST');
+        assert.strictEqual(lastRequestData.path, '/');
+        assert.strictEqual(lastRequestData.iamAuthenticate, true);
+        assert.deepStrictEqual(lastRequestData.data, expectedData);
+    });
+
+    it('should throw an error if the quota is not a strictly positive number', () => {
+        const quota = -100;
```
Comment (ghost) [inline tests/unit/updateAccountQuota.js:57]: Worth testing with 0 (means no quota - we want to use DeleteQuota for that, even if putting a very big quota may achieve the same result from the customer point of view)

**Review by ghost — APPROVED** [review]
Some minor comments

File: bin/vaultclient [hunk 9]
```diff
@@ -213,20 +213,25 @@ program
 
 program
     .command('update-account-quota')
-    .option('--account-name <NAME>', 'Name of the account')
     .option('--quota <QUOTA>', 'Maximum quota for the account', parseInt)
+    .option('--account-name <NAME>', 'Name of the account')
```
Comment (francoisferrand) [inline bin/vaultclient:218]: why this change?

File: lib/IAMClient.js [hunk 10]
```diff
@@ -284,40 +284,90 @@ class VaultClient {
...
-            'accountName is required');
-        assert(typeof quota === 'number' && !conv
-            && quota >= 0, 'Quota must be a non-negative number');
-        this.request('POST', '/', true, callback, {
+    updateAccountQuota(quota, options, callback) {
+        let accountName;
+        let ci;
+        if (options !== undefined && Object.keys(options).length !== 0) {
+            ({ accountName, ci } = options);
+        }
```
Comment (francoisferrand) [inline lib/IAMClient.js]:
```suggestion
        const { accountName, ci } = options || {};
```

File: lib/IAMClient.js [hunk 11]
```diff
@@ -284,40 +284,90 @@ class VaultClient {
...
+        const data = {
             Action: 'UpdateAccountQuota',
             Version: '2010-05-08',
-            AccountName: accountName,
             quotaMax: quota,
-        });
+        };
+        if (ci) {
```
Comment (francoisferrand) [inline lib/IAMClient.js]: should introduce a test flag in the API of the component: esp. here what this really does is simply skip validation of quota field & type of accoujnt name....

File: bin/vaultclient (hunk 9 above)
Comment (benzekrimaha) [reply bin/vaultclient:218]: this was added to enable the requester to give an account name it is not mandatory rather optional , cf design part : 
image

File: tests/unit/updateAccountQuota.js (hunk 8 above)
Comment (benzekrimaha) [reply tests/unit/updateAccountQuota.js:57]: done here : https://github.com/scality/vaultclient/pull/442/commits/868432058be9dfaddab0722ab648187cecc4b881#diff-46b3eb00304ed9af9e8497a7365f0763dab8993d18a9ae8427f7d590ba05803eR63-R71

File: lib/IAMClient.js (hunk 10 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: done here : https://github.com/scality/vaultclient/pull/442/commits/868432058be9dfaddab0722ab648187cecc4b881#diff-86a8509a2ad22b19f93123bda4efc46cfa71b4c2bbfded93e5f6e1a25e69f180R295

File: lib/IAMClient.js (hunk 7 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: done here : https://github.com/scality/vaultclient/pull/442/commits/868432058be9dfaddab0722ab648187cecc4b881#diff-86a8509a2ad22b19f93123bda4efc46cfa71b4c2bbfded93e5f6e1a25e69f180R321-R323

File: lib/IAMClient.js (hunk 11 above)
Comment (ghost) [reply lib/IAMClient.js]: The point of this change is not to alter the API in any way; the API performs validation of the inputs in any case, and we want to test that. However historically vaultclient comes with a bunch of pre-validation of the requests, so that we avoid sending malformed requests. This is similar to what we see in aws cli.
And like aws cli, this introduces an option (could be a better name than `ci`), that disable input validation on vaultclient's side. This is similar to the `parameter_validation` aws cli parameter. This allows us to test the API-side validation, in a functional way

Comment (francoisferrand) [reply lib/IAMClient.js]: > This allows us to test the API-side validation, in a functional way

api-side validation can be tested anyway by calling the api directly (bypassing vaultclient), even if vaultclient is used in the test for setup/...
"Functionally", vaultclient is expected to do validation : so testing the validation of the API using the tool in a tweaked mode is not really the functional case...

> And like aws cli, this introduces an option (could be a better name than ci), that disable input validation on vaultclient's side. This is similar to the parameter_validation aws cli parameter. 

The parameter is part of the API, and introduces complexity in the code : so the real question is really "is such a flag really useful", e.g. is it worth the added maintenance effort.
→ if not or if in doubt, let's just remove it, keep things simple and code lean
→ if this is really a useful feature, it MUST be renamed and define a proper semantics for the flag (like `parameter_validation`, which has a very clear semantic)

File: lib/IAMClient.js [hunk 12]
```diff
@@ -284,40 +284,76 @@ class VaultClient {
...
+        }
+        this.request('POST', '/', true, callback, data);
     }
 
     /**
      * Delete Quota of an account
      *
-     * @param {string} accountName - account name
+     * @param {object} options - additional creation params
+     * @param {string} [options.accountName] - account name
```
Comment (francoisferrand) [inline lib/IAMClient.js]: this will break compatibility:
* is it really acceptable?
* if we break compatibility, we should create a new major version. Otherwise, we need to maintain compatibility (possibly supporting both accountName and options.accountName)
* do we expect to have more parameters? if not, why change the api at all...

File: lib/IAMClient.js [hunk 13]
```diff
@@ -284,40 +284,76 @@ class VaultClient {
...
+     *  set to true, it helps sending invalid requests, to ensure
+     *  input validation.
```
Comment (francoisferrand) [inline lib/IAMClient.js]: leftover from `ci` param

File: lib/IAMClient.js [hunk 14]
```diff
@@ -284,40 +284,76 @@ class VaultClient {
...
-            'accountName is required');
-        this.request('POST', '/', true, callback, {
+    deleteAccountQuota(options, callback) {
+        const data = {
             Action: 'DeleteAccountQuota',
             Version: '2010-05-08',
-            AccountName: accountName,
-        });
+        };
+        if (options !== undefined && Object.keys(options).length !== 0) {
```
Comment (francoisferrand) [inline lib/IAMClient.js]: this means that if no options are passed, the call with succeed : is `accountName` really optional? does it make sense to call `deleteAccountQuota` without specifying the account?

File: lib/IAMClient.js [hunk 15]
```diff
@@ -284,40 +284,76 @@ class VaultClient {
...
+
+    /**
+     * Get Quota of an account
+     *
+     * @param {object} options - additional creation params
+     * @param {string} [options.accountName] - account name
+     * @param {VaultClient~requestCallback} callback - callback
+     * @returns {undefined}
+     */
+    getAccountQuota(options, callback) {
```
Comment (francoisferrand) [inline lib/IAMClient.js]: * same thing about the "optional" accountName
* should follow the same API style as the other api: if we stick to `accountName` (not `options`) to keep compatibility, this api should be updated to match...

File: lib/IAMClient.js [hunk 16]
```diff
@@ -284,40 +284,76 @@ class VaultClient {
     /**
      * Update Quota of an account
      *
-     * @param {string} accountName - account name
      * @param {number} quota - maximum quota for the account
+     * @param {object} options - additional creation params
+     * @param {string} [options.accountName] - account name
```
Comment (francoisferrand) [inline lib/IAMClient.js]: breaks compatibility, see below

File: lib/IAMClient.js (hunk 15 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: the compatibility has been restored , here : https://github.com/scality/vaultclient/pull/442/commits/bedefa240fb8f76860557af58533008a8b8c0bb1#diff-6ed3f39879bec9944e1a6e9d0d3c1bcbc7209a7eab826e5c3819048e4841163dR230

File: lib/IAMClient.js (hunk 16 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: done here : https://github.com/scality/vaultclient/pull/442/files#diff-6ed3f39879bec9944e1a6e9d0d3c1bcbc7209a7eab826e5c3819048e4841163dR222

File: lib/IAMClient.js (hunk 14 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: this has been removed , now we do check directly if the accountName is passed to the function see here : https://github.com/scality/vaultclient/pull/442/files#diff-86a8509a2ad22b19f93123bda4efc46cfa71b4c2bbfded93e5f6e1a25e69f180R330

File: lib/IAMClient.js (hunk 12 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: we're no longer breaking the compatibility see commit : https://github.com/scality/vaultclient/pull/442/commits/bedefa240fb8f76860557af58533008a8b8c0bb1#diff-6ed3f39879bec9944e1a6e9d0d3c1bcbc7209a7eab826e5c3819048e4841163d

File: lib/IAMClient.js (hunk 11 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: the flag has been renamed disableValidation see here : https://github.com/scality/vaultclient/pull/442/files#diff-845cfb35dcf0685889460bfb3868ad667256bef6c8c053f7d12a06987297e75dR20-R35

Comment (francoisferrand) [reply lib/IAMClient.js]: If we really want to this way, I think we should stick to `parameter_validation` (or `parameterValidation`), as it is clearer and similar to AWS

File: lib/IAMClient.js [hunk 17]
```diff
@@ -290,17 +292,25 @@ class VaultClient {
...
+                data.AccountName = accountName;
+            }
+        } else {
+            const quotaIsValid = typeof quota === 'number' && !Number.isNaN(Number.parseInt(quota, 10)) && quota > 0;
+            assert(quotaIsValid, 'Quota must be a strictly positive number');
+            if (accountName) {
+                assert(typeof accountName === 'string', 'the account name, if set, should be a string');
+                data.AccountName = accountName;
+            }
+        }
```
Comment (francoisferrand) [inline lib/IAMClient.js]:
```suggestion
          const quotaIsValid = typeof quota === 'number' && !Number.isNaN(Number.parseInt(quota, 10)) && quota > 0;
          assert(disableValidation || quotaIsValid, 'Quota must be a strictly positive number');
          if (accountName) {
              assert(disableValidation || typeof accountName === 'string', 'the account name, if set, should be a string');
              data.AccountName = accountName;
          }
        }
```

File: lib/IAMClient.js [hunk 18]
```diff
@@ -311,13 +321,42 @@ class VaultClient {
...
-            'accountName is required');
-        this.request('POST', '/', true, callback, {
+        const data = {
             Action: 'DeleteAccountQuota',
             Version: '2010-05-08',
-            AccountName: accountName,
-        });
+        };
+
+        const disableValidation = this.disableValidation;
```
Comment (francoisferrand) [inline lib/IAMClient.js]: why alias the variable?

File: lib/IAMClient.js [hunk 19]
```diff
@@ -290,17 +292,25 @@ class VaultClient {
      * @returns {undefined}
      */
     updateAccountQuota(accountName, quota, callback) {
-        const conv = Number.isNaN(Number.parseInt(quota, 10));
-        assert(typeof accountName === 'string' && accountName !== '',
-            'accountName is required');
-        assert(typeof quota === 'number' && !conv
-            && quota >= 0, 'Quota must be a non-negative number');
-        this.request('POST', '/', true, callback, {
+        const disableValidation = this.disableValidation;
```
Comment (francoisferrand) [inline lib/IAMClient.js]: why alias the variable?

File: lib/IAMClient.js (hunk 17 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: actually with this change => whenever the quota is not valid the error will be thrown regardless of the parameter (even if we don't want to check that it is valid or not which is the goal behind the flag)

File: lib/IAMClient.js (hunk 19 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: fixed here : https://github.com/scality/vaultclient/pull/442/commits/29e09a955ef3aef0716b15f9f096a2739e6f0c04

File: lib/IAMClient.js (hunk 18 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: fixed here : https://github.com/scality/vaultclient/pull/442/commits/29e09a955ef3aef0716b15f9f096a2739e6f0c04

File: lib/IAMClient.js (hunk 11 above)
Comment (benzekrimaha) [reply lib/IAMClient.js]: flag name changed here : https://github.com/scality/vaultclient/pull/442/commits/8add00de87cae735118d30f71bb56465d4aac546#diff-6ed3f39879bec9944e1a6e9d0d3c1bcbc7209a7eab826e5c3819048e4841163dR92

File: bin/vaultclient [hunk 20]
```diff
@@ -213,20 +215,28 @@ program
 
 program
     .command('update-account-quota')
-    .option('--account-name <NAME>', 'Name of the account')
+    .option('--account-name [NAME]', 'Name of the account')
     .option('--quota <QUOTA>', 'Maximum quota for the account', parseInt)
+    .option('--parameter-validation', 'Disable validation of quota')
```
Comment (anurag4DSB) [inline bin/vaultclient:220]: In which cases will we need disabling validation?

Comment (benzekrimaha) [reply bin/vaultclient:220]: explained here : https://github.com/scality/vaultclient/pull/442#discussion_r1571195119

File: bin/vaultclient [hunk 21]
```diff
@@ -213,20 +215,28 @@ program
...
     .action(action.bind(null, 'update-account-quota', (client, args) => {
-        client.updateAccountQuota(args.accountName, args.quota,
-            handleVaultResponse);
+        client.updateAccountQuota(args.accountName ,args.quota, handleVaultResponse);
```
Comment (williamlardier) [inline bin/vaultclient:222]: This change is not strictly needed

File: lib/IAMClient.js [hunk 22]
```diff
@@ -76,6 +77,7 @@ class VaultClient {
         this.log = new this.logApi.Logger('VaultClient');
         this._path = path;
         this.useAuthenticatedAdminRoutes = false;
+        this.parameterValidation = parameterValidation !== undefined ? parameterValidation : true;
```
Comment (ghost) [inline lib/IAMClient.js]: If we want to enforce a boolean here
```suggestion
        this.parameterValidation = parameterValidation !== undefined ? !!parameterValidation : true;
```

File: lib/IAMClient.js [hunk 23]
```diff
@@ -311,13 +315,38 @@ class VaultClient {
...
+     */
+    getAccountQuota(accountName, callback) {
+        const data = {
+            Action: 'GetAccountQuota',
+            Version: '2010-05-08',
+        };
+
+        if (accountName) {
+            assert(!this.parameterValidation || typeof accountName === 'string',
+            'the account name, if set, should be a string');
```
Comment (ghost) [inline lib/IAMClient.js]:
```suggestion
            assert(!this.parameterValidation || typeof accountName === 'string',
                'the account name, if set, should be a string');
```

File: lib/IAMClient.js [hunk 24]
```diff
@@ -290,17 +292,19 @@ class VaultClient {
...
             Version: '2010-05-08',
-            AccountName: accountName,
             quotaMax: quota,
-        });
+        };
+        const quotaIsValid = typeof quota === 'number' && !Number.isNaN(Number.parseInt(quota, 10)) && quota > 0;
+          assert(!this.parameterValidation || quotaIsValid, 'Quota must be a strictly positive number');
+          if (accountName) {
+              assert(!this.parameterValidation || typeof accountName === 'string',
+                'the account name, if set, should be a string');
```
Comment (ghost) [inline lib/IAMClient.js]:
```suggestion
              assert(!this.parameterValidation || typeof accountName === 'string',
                  'the account name, if set, should be a string');
```

**Review by ghost — APPROVED** [review]
Some minor comments, otherwise lgtm

File: bin/vaultclient [hunk 25]
```diff
@@ -213,20 +215,28 @@ program
 
 program
     .command('update-account-quota')
-    .option('--account-name <NAME>', 'Name of the account')
+    .option('--account-name [NAME]', 'Name of the account')
```
Comment (francoisferrand) [inline bin/vaultclient]: nit: keeping `<NAME>` was better IMO, the brackets are more often used for alternative or lists...
In any case, should be consistent with next line, where `<QUOTA>` is used (and be consistent in every command docs...)

File: bin/vaultclient (hunk 21 above)
Comment (francoisferrand) [inline bin/vaultclient]:
```suggestion
        client.updateAccountQuota(args.accountName, args.quota, handleVaultResponse);
```

File: lib/IAMClient.js [hunk 26]
```diff
@@ -290,17 +292,19 @@ class VaultClient {
...
             Action: 'UpdateAccountQuota',
             Version: '2010-05-08',
-            AccountName: accountName,
             quotaMax: quota,
-        });
+        };
+        const quotaIsValid = typeof quota === 'number' && !Number.isNaN(Number.parseInt(quota, 10)) && quota > 0;
+          assert(!this.parameterValidation || quotaIsValid, 'Quota must be a strictly positive number');
+          if (accountName) {
+              assert(!this.parameterValidation || typeof accountName === 'string',
```
Comment (francoisferrand) [inline lib/IAMClient.js]: indent:
```suggestion
        assert(!this.parameterValidation || quotaIsValid, 'Quota must be a strictly positive number');
        if (accountName) {
            assert(!this.parameterValidation || typeof accountName === 'string',
```

File: lib/IAMClient.js [hunk 27]
```diff
@@ -290,17 +292,19 @@ class VaultClient {
...
+                  'the account name, if set, should be a string');
+              data.AccountName = accountName;
+          }
```
Comment (francoisferrand) [inline lib/IAMClient.js]: indent:
```suggestion
        }
```

File: bin/vaultclient (hunk 25 above)
Comment (benzekrimaha) [reply bin/vaultclient]: done here: https://github.com/scality/vaultclient/pull/442/files#diff-6ed3f39879bec9944e1a6e9d0d3c1bcbc7209a7eab826e5c3819048e4841163dR218

File: bin/vaultclient (hunk 20 above)
Comment (anurag4DSB) [reply bin/vaultclient:220]: Thanks for the context. THats not very easy to find in a resolved conversation. Ideally this info should be mentioned in either the commit description or commit story.
As we are validating options sent to vaultclient I would suggest `validateOptions`.

File: lib/IAMClient.js [hunk 28]
```diff
@@ -290,17 +292,19 @@ class VaultClient {
...
-            AccountName: accountName,
             quotaMax: quota,
-        });
+        };
+        const quotaIsValid = typeof quota === 'number' && !Number.isNaN(Number.parseInt(quota, 10)) && quota > 0;
+        assert(!this.parameterValidation || quotaIsValid, 'Quota must be a strictly positive number');
+        if (accountName) {
+            assert(!this.parameterValidation || typeof accountName === 'string',
+                  'the account name, if set, should be a string');
+              data.AccountName = accountName;
```
Comment (francoisferrand) [inline lib/IAMClient.js]:
```suggestion
            data.AccountName = accountName;
```

Comment (ghost) [reply lib/IAMClient.js]: Done and squashed (I'm helping @benzekrimaha to optimize time)

**Review by francoisferrand — APPROVED** [review]
